### PR TITLE
Full node support for Dash

### PR DESCRIPTION
**Pull request by dev-warrior777**
Full node support for Dash

[original pr](https://github.com/decred/dcrdex/pull/2380)

### CODE REVIEW

**Comment by JoeGruffins** [comment]
Why the new pr?

**Comment by dev-warrior777** [comment]
> Why the new pr?

My fork git repo git was messed up ..

**Review by JoeGruffins — APPROVED** [review]
I'm still not sure about the Firo ports change, but everything else looks good to me.

I still see the error, when applying #2425 will see `[ERR] CORE[dash]: error unlocking spent coins: rawrequest (lockunspent) error: -8: Invalid parameter, expected unspent output`

It looks to me like the output is already spent from the point of view of the node before we call unlock. While we should figure out what's going on there, it doesn't look like a real problem to me, just a difference between bitcoind and dashd. The coins should be spent by that time by the split tx, but bitcoind does not remove them from the locked list automatically while dashd does maybe? Not sure.

**Comment by dev-warrior777** [comment]
> I'm still not sure about the Firo ports change,

[Firo RPC Ports](https://github.com/firoorg/firo/blob/079d57e7a0aaee20242f3470e491171016546bb4/src/chainparamsbase.cpp#L37)

8888 .. Firo is Asian and 8888 is prosperity ;-)

**Comment by dev-warrior777** [comment]
> I still see the error, when applying #2425 `[ERR] CORE[dash]: error unlocking spent coins: rawrequest (lockunspent) error: -8: Invalid parameter, expected unspent output`

I went into this deeper: In some cases I do actually see 'wrong' coins returned by the logic occasionally. Meaning they are not in the `listlockedunspent` list as far as Dash is concerned when returned. And this is the Dash RPC error. Since this was with debugger breakpoint not easy to tell if something not locked and a race was lost

File: client/asset/dash/regtest/recycled-addrs.txt
Comment (buck54321) [inline client/asset/dash/regtest/recycled-addrs.txt:1]: Hmm. I'm not sure how this file ended up here. You can drop this file and its containing directory.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/asset/dash/regtest/recycled-addrs.txt
Comment (dev-warrior777) [reply client/asset/dash/regtest/recycled-addrs.txt:1]: Done

This is a side effect of [this pr](https://github.com/decred/dcrdex/pull/2368) and directly running Dex code that asks for an address from inside the git code tree. In this case `client/asset/dash/regnet_test.go`.

I usually remember to clean up and did put in a pr to hide these directories from git but the granularity of the `.gitignore` is kinda unknown .. is it any code? I certainly see with other `client asset` tests.

Perhaps we should at least `.gitignore` any `regtest/` directory above a certain height in the code tree. But what should be that height is not obvious.

Comment (buck54321) [reply client/asset/dash/regtest/recycled-addrs.txt:1]: > This is a side effect of [this pr](https://github.com/decred/dcrdex/pull/2368) and directly running Dex code that asks for an address from inside the git code tree. In this case `client/asset/dash/regnet_test.go`.

That file doesn't seem to exist. You can give steps to reproduce these misplaced files?

Comment (dev-warrior777) [reply client/asset/dash/regtest/recycled-addrs.txt:1]: ### Sequence

- Start Firo harness 

```bash
cd dex/testing/firo
#export PATH to firod/firo-cli 
./harness.sh
```
- run asset tests

```bash
cd client/asset/firo
# delete any regtest dir if found
go test -v -count=1 -tags=harness ./...
```
You should see a new directory `regtest` if the livetest code was successful

Comment (dev-warrior777) [reply client/asset/dash/regtest/recycled-addrs.txt:1]:
```bash
dev@dev-ThinkPad-T430 ~/dex/dcrdex/client/asset/firo (harness_mining)$ ls
firo.go  regnet_test.go
dev@dev-ThinkPad-T430 ~/dex/dcrdex/client/asset/firo (harness_mining)$ go test -v -count=1 -tags=harness ./...
=== RUN   TestWallet
    livetest.go:197: Setting up alpha/beta/gamma wallet backends...
2023-07-21 00:56:01.266 [DBG] TEST[alpha.alpha][RPC]: Node at 127.0.0.1:53768 reports subversion "/Satoshi:0.14.12.1/"
2023-07-21 00:56:01.302 [INF] TEST[alpha.alpha]: Connected wallet with current best block 3dc04a8987a951ae30bf7ce89653ae873ef1fd4d12b219b7160080b39bbd7d1d (352)
peer count = 3, err = <nil>2023-07-21 00:56:01.303 [DBG] TEST[beta.beta][RPC]: Node at 127.0.0.1:53769 reports subversion "/Satoshi:0.14.12.1/"

. . .

rate = 100, actual fee rate = 100 (22500 for 225 bytes), change = true
2023-07-21 00:56:04.972 [INF] TEST: Withdrew with a9a9f48aa475e4482be4e34919eb8370b57915b2f27940eb69a63550314ad421:0
--- PASS: TestWallet (3.71s)
=== RUN   TestFetchExternalFee
#### External fee rate fetched:: 1 sat/B
--- PASS: TestFetchExternalFee (0.91s)
PASS
ok  	decred.org/dcrdex/client/asset/firo	4.632s
dev@dev-ThinkPad-T430 ~/dex/dcrdex/client/asset/firo (harness_mining)$ ls
firo.go  regnet_test.go  regtest
dev@dev-ThinkPad-T430 ~/dex/dcrdex/client/asset/firo (harness_mining)$ 

```

Comment (dev-warrior777) [reply client/asset/dash/regtest/recycled-addrs.txt:1]: image

**Comment by chappjc** [comment]
Remember not to sync with master by merging into your feature branch.  Please rebase.  You  may squash away all but one commit at this point, and that might be necessary to eliminate the merge commit.

**Review by buck54321 — APPROVED** [review]

File: dex/networks/dash/params.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	h, err := chainhash.NewHashFromStr(hash)
+	if err != nil {
+		panic(err.Error())
+	}
+	return h
+}
+
+const (
+	DefaultFee          = 1  // 1 sat/byte
+	DefaultFeeRateLimit = 20 // 20 sats/byte
```
[the remark this answers is not in the document]
Comment (dev-warrior777) [reply dex/networks/dash/params.go:23]: - 1 sat/byte? Dash is considered a low fee/cost chain. I actually did not see higher than this on blockchair.

- 20 sats/byte .. I am not sure but will ask Joel at Dash to ask pasta or another dev and I will read the C++ code again

File: dex/testing/dash/harness.sh [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+CLI="dash-cli"
+RPC_USER="user"
+RPC_PASS="pass"
+ALPHA_LISTEN_PORT="20775"
+BETA_LISTEN_PORT="20776"
+ALPHA_RPC_PORT="20777"
+BETA_RPC_PORT="20778"
+# additional daemon args - currently none
+EXTRA_ARGS=
+#EXTRA_ARGS="-usehd=1"
```
[the remark this answers is not in the document]
Comment (dev-warrior777) [reply dex/testing/dash/harness.sh:23]: I can delete the commented out args example but would like to keep the empty parameter `EXTRA_ARGS=` as the preferred way to add more unusual args in future.

File: dex/networks/dash/params.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dev-warrior777) [reply dex/networks/dash/params.go:23]: [Dash wallet fees](https://github.com/dashpay/dash/blob/a884f0c53637e850066a7a31b7f3825e40195e8c/src/wallet/wallet.h#L68)

`COIN = 100000000`

**Pull request merged**